vm/devices/storage/disklayer_sqlite/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+use guestmem::MemoryRead;
+use guestmem::MemoryWrite;
+use inspect::Inspect;
+use parking_lot::Mutex;
+use rusqlite::Connection;
+use scsi_buffers::RequestBuffers;
+use std::sync::Arc;
+
+/// Formatting parameters provided to [`FormatOnAttachSqliteDiskLayer::new`].
+///
+/// Optional parameters which are not provided will be determined by reading the
+/// metadata of the layer being attached to.
+#[derive(Inspect, Copy, Clone)]
+pub struct IncompleteFormatParams {
+    /// Should the layer be considered logically read only (i.e: a cache layer)
+    pub logically_read_only: bool,
+    /// The size of the layer in bytes.
+    pub len: Option<u64>,
+}
+
+/// Formatting parameters provided to [`SqliteDiskLayer::new`]
+#[derive(Inspect, Copy, Clone)]
+pub struct FormatParams {
+    /// Should the layer be considered logically read only (i.e: a cache layer)
+    pub logically_read_only: bool,
+    /// The size of the layer in bytes. Must be divisible by `sector_size`.
+    pub len: u64,
+    /// The size of each sector.
+    pub sector_size: u32,
+}
+
+/// A disk layer backed by sqlite, which lazily infers its topology from the
+/// layer it is being stacked on-top of.
+#[derive(Inspect)]
+#[non_exhaustive]
+pub struct FormatOnAttachSqliteDiskLayer {
+    dbhd_path: String,
+    read_only: bool,
+    format_dbhd: IncompleteFormatParams,
+}
+
+impl FormatOnAttachSqliteDiskLayer {
+    /// Create a new sqlite-backed disk layer, which is formatted when it is
+    /// attached.
+    pub fn new(
+        dbhd_path: String,
+        read_only: bool,
+        format_dbhd: IncompleteFormatParams,
+    ) -> anyhow::Result<Self> {
+        Ok(Self {
+            dbhd_path,
+            read_only,
+            format_dbhd,
+        })
+    }
+}
+
+/// A disk layer backed entirely by sqlite.
+#[derive(Inspect)]
+pub struct SqliteDiskLayer {
+    #[inspect(skip)]
+    conn: Arc<Mutex<Connection>>,
+    meta: schema::DiskMeta,
+    read_only: bool,
+}
+
+impl SqliteDiskLayer {
+    /// Create a new sqlite-backed disk layer.
+    pub fn new(
+        dbhd_path: String,
+        read_only: bool,
+        format_dbhd: Option<FormatParams>,
+    ) -> anyhow::Result<Self> {
+        // DEVNOTE: sqlite _really_ want to be in control of opening the file,
+        // since it also wants to read/write to the runtime "sidecar" files that
+        // get created when accessing the DB (i.e: the `*-shm` and `*-wal`
+        // files)
+        //
+        // This will make it tricky to sandbox SQLite in the future...
+        //
+        // One idea: maybe we could implement a small SQLite `vfs` shim that
+        // lets use pre-open those particular files on the caller side, and hand
+        // them to sqlite when requested (vs. having it `open()` them itself?)
+        let conn = Connection::open_with_flags(dbhd_path, {
+            use rusqlite::OpenFlags;
+
+            let mut flags = OpenFlags::SQLITE_OPEN_NO_MUTEX | OpenFlags::SQLITE_OPEN_URI;
```

Comment:
Why `SQLITE_OPEN_URI`?

Reply:
This is based off the implementation of `OpenFlags::default()`. Seems harmless enough to leave in?

Reply:
I don't know if we want to allow people to pass in URIs, though. It seems a bit too powerful. <https://www.sqlite.org/uri.html>

Reply:
Fair enough! And I suppose it won't really matter if/when we get around to some of the VFS stuff. I'll go ahead and remove it...